### ISSUE TITLE
Seed/Sample for node package, node app consuming other Visual Studio project.

### ISSUE THREAD

**Issue by fforjan** [Microsoft/nodejstools#1294]
Hey guys, 

Not sure where to ask this question, so here is it ...
I've currently hacked a simple nodejs visual studio project which:
 - consume our internal .net assemblies
 - wrapped it into a js object via edgejs
 - consume it into a node application
this node app is working fine.

In order to go the next step, I need some guidance from you.
1. I want to create a new nodejs visual studio project which in fact generate the output for a node package.
This project should be using typescript so I"m expecting there is 2 outputs, the node package itself and the d.ts.
Do you have any sample/seed for this  ?

2. The output node package should include our internal assemblies.  The nodejs visual studio project is existing within the same solution as our C# project. How to configure this project so that it is build after our C# project so I can use a pre-build/post-build step to deploy the dll at the right place ?

3. From a build agent, is there a simply way to publish this node package somewhere - network share as first step would be awesome, we're looking into artifactory for our nuget/node repository. This repository should be private to our company and we don't want to have this available on internet for now. (the idea behind is other team / team member could create their node application consuming this node package)

3. Now I also want to create a new nodejs visual studio project which is my node application, consider it as a sample application. This application should consume the node package generated by the previous project if it was build by a developer or the build agent.
Again, seed/sample would be welcome.

PS: I'm pretty sure I can do something with gulp, but it seems maybe too much of a hack  ?

**Comment by mjbvz** [comment]
I don't know of any good examples that show this setup. This is not something we support directly but it all sounds possible, although perhaps rather complicated. I don't know all the details of your project but here are some questions and thoughts:

1. Are you targeting a continuous integration on something like VSTS, or targeting local builds?

1. When you say generate a node package, do you mean just generating the `js` files from the `ts` files? Or do you mean outputting files into package folder and generating a `package.json` for them?

  The first case is already supported, and the second case should be possible. You'd need some custom MSBuild targets if you really want to automate the entire package generation flow. 

1. For local publishing, you can either setup a local instance of NPM and publish to that, or just copy the package folder to a network share and npm install from there directly: `npm install //path/to/package/on/share`. How you copy the files is up to you. There are msbuild targets like `Copy` that may help, or you can run any command line script with `Exec`.

1. As for actually generating an NTVS `njsproj` file, NTVS doesn't provide much help for this, but you can use whatever tools you want to accomplish this. The project file is just xml, so you can start with a template file and fill in whatever generated values you need. Articles like this may provide a some ideas: https://msdn.microsoft.com/en-us/library/ee847423.aspx

Hope that provides some starting points.

**Comment by fforjan** [comment]
1. well, we are targeting both ?

Developer experience : when the developer build on its machine, it should work.
Build machine : when the build is triggered, we generates the right package

2. Generating the output files for the package folder.
What's your recommendation for the .d.ts ? because we want to distribute the js file only but with the d.ts so that other teams can also use typescript.

3.  thanks ! I didn't now the install, will try it ! No need of msbuild targets, workflow is that none of developers produce a product artifacts - nuget or nodejs package, only build machine can do this.

4. Maybe I wasn't clear, but now I got one project generated the package content, how do I create a second project consuming the package. 
But from what you're saying is, I should be able to use the Npm install command with file path... let's see how this is working :)

**Comment by mjbvz** [comment]
\1. I think you'll mainly want to stick to msbuild logic then. VSTS provides some helper tasks for things like publishing in continuous integration, but I don't know if these can be used for local builds.

\2. Just include the `*.d.ts` file in the package, probably along with the `*.ts` files and any source maps. The Typescript project has a resource on this: http://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html

\4. If the second project is a node project, yes the easiest approach is to npm install from the path. If the share requires authentication, some additional work may be needed . You can always copy the file locally using another tool and then install from there.

Let me know how that goes.

**Comment by fforjan** [comment]
1. msbuild is relying on project references for the build order. I don't think there is any UI to do this, but do you know if I can hack the project file to support project references ? did you ever try ?

2. Awesome. Still lot of things to learn on publishing it seems.

Will try to change a sample with this working :)

**Comment by fforjan** [comment]
You can have a look here : https://github.com/fforjan/NodeJSSOlution
(application is not working, didn't spend time to make it work, but you got the idea...)

But : 
1 - how to ensure that projects are build in the right order ?
2 - If a developer do a modification on the tslib, we should clear the nodejs module into the app ?

Maybe I would rephrase, what's the expected workflow in that case when you got multiple projects ? Is there something I'm doing in the wrong way and/or should be done differently ?

**Comment by fforjan** [comment]
@mjbvz any idea how to deals with project build order in the sample ? And also how to force reinstalling the npm package ?

**Comment by fforjan** [comment]
@mjbvz Still no feedback ? That's blocking me from deploying it to the dev team as I don't have a working workflow.

**Comment by mjbvz** [comment]
Thank you for taking the time to put together the sample. I took a look and have a few ideas for you to investigate further:

You can configure build order and build dependencies by right clicking on the target project in the solution explorer and selecting `Build Dependencies -> Project Dependencies` from the context menu. After adding a dependency from `MyDemoApp` to `MyTSLib`, I see the lib is built before the app is. 

This is not the root problem however.

In `MyDemoApp`, you have `my-tslib` listed as a local dependency:

```json
 "dependencies": {
    "my-tslib": "..\\MyTSLib"
  }
```

That means that when you `npm install`, npm installs `my-tslib`  from the local path. It does not keep the local copy of `my-tslib` inside `MyDemoApp/node_modules` up to date after that.

One way to fix this is by telling `require` where the find this module. The dependency entry in the `package.json` only specifies where to install from, not how modules are resolved with `require`. To pick up the local version of `MyTslib` in your app, change:

```ts
import MyLIb from "my-tslib";
```

to:

```ts
import MyLIb from "../MyTslib";
```

If you want to keep the application code unchanged, I believe you'll want to use [`npm link`](https://docs.npmjs.com/cli/link). That ensures that node picks up the current code from `my-tslib` when it is required. After making that change, I see the project working as expected. 

Let me know if that works.

**Comment by fforjan** [comment]
Thanks for your answer !
By experiences, project dependencies setup in Visual Studio are not honored by command line tools, including MsBuild - https://github.com/Microsoft/msbuild/issues/421.
So it means it may work for developers but more likely not for a build system.
I've tried adding project reference xml node into the project file but I can't see that information into visual studio and I don't know how to display msbuild build order in command line.

npm link may work on a developer workstation but would require manual steps and would not support multiple branches - I didn't try but that's my feeling with the documentation..

It seems the only way is to remove the package approach and use the local import.
The only drawback is that we can't just publish the source code of this sample app like this, we will need to patch it before publishing it, and I'm not a fan of this.

**Comment by mjbvz** [comment]
Take a look at these articles too regarding project references: 

https://blogs.msdn.microsoft.com/visualstudio/2010/12/21/incorrect-solution-build-ordering-when-using-msbuild-exe/
https://support.microsoft.com/en-us/kb/964125

As for updating the `my-tslib` package itself, you can either clear the node_modules folder (or just the part for `my-tslib`) and run `npm install` again, or run `npm install my-tslib --force`. I believe that will always copy the updated version of `my-tslib` to `node_modules`. I'm not too familiar with force however and don't know how this handles dependencies of `my-tslib` or other edge cases. Either approach can be integrated into msbuild as scripts run before the application project itself is built.

For this type of situation, I've always used npm link during development and published package versions for deployment (or by using local dependencies sometimes).

**Comment by fforjan** [comment]
So I've updated my sample with :
- using the 'local' include instead of package referencing
- manually added project references for the build order
On this last item, msbuild should be calculating the right order as `msbuild MyDemoApp.njsproj` biuld all 3 projects in the right order.
I will need to play more to ensure all use cases are covered.
